File: native/core-swift/Tests/C15tCoreTests/VendorListScopeFixtures.swift
import CryptoKit
import Foundation

@testable import C15tCore

/// What one surviving vendor claims about itself, as the web filter left it in the fixture.
///
/// ``bodyId`` is the id the surviving entry states in its own `id` field, which the key it is
/// served under may or may not agree with. The filter never consults it, so
/// ``bodyIdDiffersFromKey`` is the fixture recording a disagreement the producer planted rather
/// than a rule this target invented: a prune that keyed correctly and read bodies anyway returns
/// a different set with the same count, and only a key-by-key comparison can see that.
struct VendorSurvivalClaim {
    let bodyId: Int
    let bodyIdDiffersFromKey: Bool
    let carriesOptionalFields: Bool
    let key: Int
    let name: String
}

/// One `vendor-list-scope` fixture, read off the bytes `index.json` hashed.
///
/// ``scope`` is `nil` for the vector that declares no scope at all, `[]` for the vector that
/// declares an empty one, and the declared ids otherwise. Both no-scope shapes expect the served
/// document back, which is what ``unchanged`` records: the web's filter returned the document it
/// was handed rather than a rebuilt copy of it.
///
/// ``expectedVendorKeys`` is in web order. Swift's vendor collection is a dictionary, which has no
/// order to report, so this target compares key sets and per-key entries; the ordering claim is
/// graded where a collection carries order, which is `narrowToVendorIds` in the Kotlin core.
struct VendorListScopeVector {
    let expectedDocument: JSONValue
    let expectedVendorKeys: [Int]
    let file: String
    let id: String
    /// The served keys whose entry states a different id in its own body, ascending.
    let disagreeingKeys: [Int]
    let scope: [Int]?
    let scopeSize: Int
    let served: JSONValue
    let servedVendorKeys: [Int]
    let shape: String
    let surviving: [VendorSurvivalClaim]
    let unchanged: Bool
}

/// The only place this test target opens a `vendor-list-scope` fixture.
///
/// The vectors lane owns `native/protocol` and publishes through its `index.json`; this target
/// reads that directory and writes nothing to it. The oracle behind these files is
/// `narrowGVLToVendors` in `packages/iab/src/tcf/fetch-gvl.ts`, called by the generator rather
/// than restated, so an expectation here is the browser build's answer and not Swift's opinion
/// about it.
///
/// ``all()`` enumerates through the index rather than globbing, and proves the bytes are the ones
/// the generator hashed before parsing them. A fixture whose digest moved was edited after it was
/// written, and grading this core against that would fail Swift for somebody else's keystrokes.
enum VendorListScopeFixtures {
    /// The kind these fixtures carry, and the only kind ``all()`` returns.
    static let kind = "vendor-list-scope"

    /// `native/protocol`, resolved from this file's path because `swift test` makes no promise
    /// about the working directory.
    static let directory: URL = URL(fileURLWithPath: #filePath)
        .deletingLastPathComponent()
        .deletingLastPathComponent()
        .deletingLastPathComponent()
        .deletingLastPathComponent()
        .appendingPathComponent("protocol")

    enum LoadError: Error, CustomStringConvertible {
        case missingField(file: String, detail: String)
        case hashMismatch(file: String, wanted: String, got: String)
        case malformed(file: String, detail: String)

        var description: String {
            switch self {
            case let .missingField(file, detail):
                return "\(file): \(detail)"
            case let .hashMismatch(file, wanted, got):
                return "\(file) hashes to \(got), index.json says \(wanted): regenerate rather than comparing"
            case let .malformed(file, detail):
                return "\(file): \(detail)"
            }
        }
    }

    private struct IndexEntry: Decodable {
        let file: String
        let id: String
        let kind: String
        let sha256: String
    }

    private struct Index: Decodable {
        let fixtures: [IndexEntry]
    }

    /// Every `vendor-list-scope` fixture the index lists, in a stable order.
    static func all() throws -> [VendorListScopeVector] {
        let data = try Data(contentsOf: directory.appendingPathComponent("index.json"))
        let index = try JSONDecoder().decode(Index.self, from: data)
        let listed = index.fixtures
            .filter { $0.kind == kind }
            .sorted { $0.file < $1.file }
        let vectors = try listed.map(load)

        // A file on disk the index never listed would be invisible to the loop above, which is
        // how a published vector quietly stops being graded.
        let onDisk = (try? FileManager.default.contentsOfDirectory(atPath: directory.path))?
            .filter { $0.hasPrefix("\(kind)-") && $0.hasSuffix(".json") } ?? []
        let unlisted = onDisk.filter { name in !listed.contains { $0.file == name } }.sorted()
        if !unlisted.isEmpty {
            throw LoadError.malformed(
                file: "index.json",
                detail: "holds \(unlisted.joined(separator: ", ")) on disk that it does not list"
            )
        }
        return vectors
    }

    /// The ids of every vector, so the kernel runner can prove it claims these too.
    static func ids() throws -> [String] { try all().map(\.id).sorted() }

    /// Read a fixture and prove the bytes are the ones the generator hashed.
    private static func load(_ entry: IndexEntry) throws -> VendorListScopeVector {
        let data = try Data(contentsOf: directory.appendingPathComponent(entry.file))
        let digest = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        guard digest == entry.sha256 else {
            throw LoadError.hashMismatch(file: entry.file, wanted: entry.sha256, got: digest)
        }
        guard let root = C15tJSON.parse(data) else {
            throw LoadError.malformed(file: entry.file, detail: "not JSON")
        }
        return try parse(root, entry: entry)
    }

    private static func parse(_ root: JSONValue, entry: IndexEntry) throws -> VendorListScopeVector {
        // `JSONValue` stores whole numbers as `Int64`, and every id in a vendor list is an `Int`
        // on both platforms. Reading through these two keeps a silently narrowing cast out of the
        // file: a number outside `Int` is a malformed fixture, not an id.
        guard let input = root["input"], let expected = root["expected"] else {
            throw LoadError.malformed(file: entry.file, detail: "no input or expected")
        }
        guard let served = input["document"], let expectedDocument = expected["document"] else {
            throw LoadError.malformed(file: entry.file, detail: "no input.document or expected.document")
        }
        guard let protocolVersion = root["protocolVersion"]?.intValue, protocolVersion == 1 else {
            throw LoadError.malformed(file: entry.file, detail: "protocolVersion is not this build's 1")
        }
        guard let scopeSize = id(input["scopeSize"]) else {
            throw LoadError.malformed(file: entry.file, detail: "no input.scopeSize")
        }
        guard let expectedVendorKeys = keyList(expected["vendorKeys"]) else {
            throw LoadError.malformed(file: entry.file, detail: "expected.vendorKeys is not a list of document keys")
        }
        guard let rawScope = input["vendorIds"] else {
            throw LoadError.malformed(file: entry.file, detail: "input.vendorIds is missing; the absent case is a JSON null, not an absent key")
        }
        let scope: [Int]?
        if rawScope.isNull {
            scope = nil
        } else {
            guard let ids = idList(rawScope) else {
                throw LoadError.malformed(file: entry.file, detail: "input.vendorIds is neither null nor a list of ids")
            }
            scope = ids
        }
        guard let transcript = expected["vendors"]?.arrayValue else {
            throw LoadError.malformed(file: entry.file, detail: "no expected.vendors transcript")
        }
        let surviving = try transcript.map { row -> VendorSurvivalClaim in
            guard let key = id(row["key"]), let name = row["name"]?.stringValue,
                  let bodyId = id(row["bodyId"]),
                  let differs = row["bodyIdDiffersFromKey"]?.boolValue,
                  let carriesOptional = row["carriesOptionalFields"]?.boolValue
            else {
                throw LoadError.malformed(file: entry.file, detail: "a surviving vendor is incomplete")
            }
            return VendorSurvivalClaim(
                bodyId: bodyId,
                bodyIdDiffersFromKey: differs,
                carriesOptionalFields: carriesOptional,
                key: key,
                name: name
            )
        }
        guard let servedVendors = served["vendors"]?.objectValue else {
            throw LoadError.malformed(file: entry.file, detail: "input.document.vendors is not a record")
        }
        guard let servedKeys = idKeys(servedVendors.keys) else {
            throw LoadError.malformed(file: entry.file, detail: "input.document.vendors has a key that is not an id")
        }

        // Which served records state a number that is not the key they live under. Read from the
        // document rather than trusted from a comment, because the planted pair is the whole
        // reason the key-authority vector can tell the two readings apart at all.
        let disagreeing = servedVendors
            .filter { name, record in
                guard let key = Int(name), let stated = id(record["id"]) else { return false }
                return key != stated
            }
            .keys
            .compactMap(Int.init)
            .sorted()

        return VendorListScopeVector(
            expectedDocument: expectedDocument,
            expectedVendorKeys: expectedVendorKeys,
            file: entry.file,
            id: entry.id,
            disagreeingKeys: disagreeing,
            scope: scope,
            scopeSize: scopeSize,
            served: served,
            servedVendorKeys: servedKeys,
            shape: root["shape"]?.stringValue ?? "",
            surviving: surviving,
            unchanged: expected["unchanged"]?.boolValue ?? false
        )
    }

    /// One id, or `nil` for anything that is not a whole number this platform can hold.
    /// Every served key as an id, ascending, or `nil` when one of them is not a number. A
    /// document whose keys are not ids is not the shape either core reads, so it stops here.
    private static func idKeys(_ keys: some Sequence<String>) -> [Int]? {
        var ids: [Int] = []
        for key in keys {
            guard let id = Int(key) else { return nil }
            ids.append(id)
        }
        return ids.sorted()
    }

    /// One id, or `nil` for anything that is not a whole number this platform can hold.
    private static func id(_ value: JSONValue?) -> Int? {
        guard let wide = value?.intValue, Int64(Int.min) <= wide, wide <= Int64(Int.max) else {
            return nil
        }
        return Int(wide)
    }

    /// A list of `vendors` keys, in published order.
    ///
    /// The web filter returns an object, so the keys it survived with come out of `Object.keys` --
    /// strings. The harness publishes them that way instead of quietly numbering them, which is
    /// what makes this list the document's own key order rather than a second answer about ids.
    /// Order matters here: the Kotlin core compares against it entry for entry.
    private static func keyList(_ value: JSONValue?) -> [Int]? {
        guard let items = value?.arrayValue else { return nil }
        var ids: [Int] = []
        ids.reserveCapacity(items.count)
        for item in items {
            guard let key = item.stringValue, let id = Int(key) else { return nil }
            ids.append(id)
        }
        return ids
    }

    /// A list where every member is an id, or `nil` when any member is not. Nothing is dropped
    /// on the floor here: a vector whose ids do not all read is a broken fixture, and grading
    /// against a partial id list would turn that into a green run.
    private static func idList(_ value: JSONValue?) -> [Int]? {
        guard let items = value?.arrayValue else { return nil }
        var ids: [Int] = []
        ids.reserveCapacity(items.count)
        for item in items {
            guard let id = id(item) else { return nil }
            ids.append(id)
        }
        return ids
    }
}

File: native/core-android/c15t-core/src/test/kotlin/com/c15t/core/tc/VendorListScopeFixtures.kt
package com.c15t.core.tc

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import java.io.File
import java.security.MessageDigest
import kotlin.test.fail

/**
 * What one surviving vendor claims about itself, as the web filter left it in the fixture.
 *
 * [bodyId] is the id the surviving entry states in its own `id` field, which the served
 * document's key may or may not agree with. The filter never consults it, so
 * [bodyIdDiffersFromKey] is the fixture recording a disagreement the producer planted rather
 * than a rule this build invented, and a runner that keys correctly and reads bodies anyway
 * produces a different set with the same count -- which is only visible if the surviving keys
 * are compared one by one.
 */
internal data class VendorSurvivalClaim(
	val bodyId: Long,
	val bodyIdDiffersFromKey: Boolean,
	val carriesOptionalFields: Boolean,
	val key: Long,
	val name: String,
)

/**
 * One `vendor-list-scope` fixture, read off the bytes `index.json` hashed.
 *
 * [scope] is `null` for the vector that declares no scope at all, an empty list for the vector
 * that declares an empty one, and the declared ids otherwise. Both "no scope" shapes expect the
 * served document back, which is what [unchanged] records: the web's filter returned the very
 * document it was handed rather than a rebuilt copy of it.
 */
internal data class VendorListScopeFixture(
	val expectedDocument: JsonObject,
	val expectedVendorKeys: List<Long>,
	val id: String,
	val scope: List<Long>?,
	val scopeSize: Int,
	val served: JsonObject,
	val servedVendorKeys: List<Long>,
	val shape: String,
	val surviving: List<VendorSurvivalClaim>,
	val unchanged: Boolean,
)

/**
 * The shared `vendor-list-scope` fixtures in `native/protocol`, loaded through one door.
 *
 * The vectors lane owns that directory and publishes through its `index.json`; this build reads
 * it and writes nothing to it. The oracle behind these files is `narrowGVLToVendors` in
 * `packages/iab/src/tcf/fetch-gvl.ts`, called by the generator rather than restated, so an
 * expectation here is the browser build's answer and not Kotlin's opinion about it.
 *
 * [load] enumerates through the index rather than globbing, and proves the bytes are the ones
 * the generator hashed before parsing them: a fixture whose digestmoved is a fixture that was
 * edited by hand after it was written, and grading a core against that would fail the core for
 * someone else's edit. A `vendor-list-scope` file on disk that the index never listed also
 * fails, so the set under test cannot shrink or grow quietly.
 */
internal object VendorListScopeFixtures {
	/** The kind these fixtures carry, and the only one [load] returns. */
	const val KIND: String = "vendor-list-scope"

	private const val INDEX_FILE = "index.json"

	private val json = Json { ignoreUnknownKeys = true }

	/**
	 * Every `vendor-list-scope` entry in `index.json`, in index order, with its bytes verified.
	 */
	fun load(): List<VendorListScopeFixture> {
		val directory = fixtureDirectory()
		val index = json.parseToJsonElement(File(directory, INDEX_FILE).readText()).jsonObject
		val listed = index["fixtures"]!!.jsonArray
			.map { it.jsonObject }
			.filter { it["kind"]?.jsonPrimitive?.content == KIND }

		val claimed = listed.map { entry ->
			val name = entry["file"]?.jsonPrimitive?.content
				?: fail("a $KIND index entry names no file: $entry")
			read(directory, name, entry["sha256"]?.jsonPrimitive?.content
				?: fail("$name: its index entry carries no sha256"))
		}

		val unlisted = directory.listFiles().orEmpty()
			.filter { it.name.startsWith("$KIND-") && it.name.endsWith(".json") }
			.map { it.name }
			.filterNot { name -> listed.any { it["file"]?.jsonPrimitive?.content == name } }
			.sorted()
		if (unlisted.isNotEmpty()) {
			fail(
				"$KIND fixtures on disk that $INDEX_FILE does not list: $unlisted. A runner that " +
					"enumerates the index would skip them silently."
			)
		}
		return claimed
	}

	/** The ids of every `vendor-list-scope` fixture, so the kernel runner can prove it claims them too. */
	fun ids(): List<String> = load().map { it.id }.sorted()

	/**
	 * Walk up from the Gradle module until `native/protocol/index.json` appears.
	 *
	 * The same walk ProtocolFixtureTest makes: `user.dir` is the module directory under the
	 * Gradle test JVM, and the protocol directory is a sibling of the whole native tree.
	 */
	private fun fixtureDirectory(): File {
		var cursor: File? = File(System.getProperty("user.dir")).absoluteFile
		while (cursor != null) {
			val candidate = File(cursor, "native/protocol/$INDEX_FILE")
			if (candidate.isFile) {
				return candidate.parentFile
			}
			cursor = cursor.parentFile
		}
		fail(
			"no native/protocol/$INDEX_FILE above ${File(System.getProperty("user.dir")).absolutePath}. " +
				"Run `bun run --cwd packages/react-native generate:fixtures`."
		)
	}

	/** Read one fixture, refusing it unless its bytes still hash to what the index recorded. */
	private fun read(directory: File, name: String, expectedDigest: String): VendorListScopeFixture {
		val file = File(directory, name)
		if (!file.isFile) {
			fail("$INDEX_FILE lists $name, which is not on disk")
		}
		val bytes = file.readBytes()
		val digest = MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }
		if (digest != expectedDigest) {
			fail(
				"$name is not the file $INDEX_FILE hashed ($digest != $expectedDigest). It was edited " +
					"after generation; regenerate rather than grading this build against it."
			)
		}
		return parse(name, json.parseToJsonElement(String(bytes, Charsets.UTF_8)).jsonObject)
	}

	private fun parse(fileName: String, root: JsonObject): VendorListScopeFixture {
		val input = root["input"]?.jsonObject ?: fail("$fileName: no input")
		val expected = root["expected"]?.jsonObject ?: fail("$fileName: no expected")
		val served = input["document"]?.jsonObject ?: fail("$fileName: no input.document")
		val expectedDocument = expected["document"]?.jsonObject ?: fail("$fileName: no expected.document")
		val protocolVersion = root["protocolVersion"]?.jsonPrimitive?.long
		if (protocolVersion != 1L) {
			fail("$fileName: protocolVersion $protocolVersion is not this build's 1")
		}
		val vendorIds = input["vendorIds"]
		val scope = when {
			vendorIds == null || vendorIds is kotlinx.serialization.json.JsonNull -> null
			else -> vendorIds.jsonArray.map { it.jsonPrimitive.long }
		}
		return VendorListScopeFixture(
			expectedDocument = expectedDocument,
			expectedVendorKeys = expected["vendorKeys"]!!.jsonArray.map { it.jsonPrimitive.long },
			id = root["id"]?.jsonPrimitive?.content ?: fail("$fileName: no id"),
			scope = scope,
			scopeSize = input["scopeSize"]?.jsonPrimitive?.int ?: fail("$fileName: no input.scopeSize"),
			served = served,
			servedVendorKeys = served["vendors"]!!.jsonObject.keys.map { it.toLong() },
			shape = root["shape"]?.jsonPrimitive?.content ?: fail("$fileName: no shape"),
			surviving = expected["vendors"]!!.jsonArray.map { row ->
				val fields = row.jsonObject
				VendorSurvivalClaim(
					bodyId = fields["bodyId"]!!.jsonPrimitive.long,
					bodyIdDiffersFromKey = fields["bodyIdDiffersFromKey"]!!.jsonPrimitive.boolean,
					carriesOptionalFields = fields["carriesOptionalFields"]!!.jsonPrimitive.boolean,
					key = fields["key"]!!.jsonPrimitive.long,
					name = fields["name"]!!.jsonPrimitive.content,
				)
			},
			unchanged = expected["unchanged"]?.jsonPrimitive?.boolean
				?: fail("$fileName: no expected.unchanged"),
		)
	}

	/** The optional halves of `gvlVendorSchema`, so "the survivor kept its own fields" is one check. */
	val OPTIONAL_VENDOR_FIELDS: List<String> = listOf(
		"dataCategories",
		"dataRetention",
		"deletedDate",
		"deviceStorageDisclosureUrl",
		"overflow",
	)

	/** The records that sit beside `vendors` in the document, which no scope may touch. */
	val SIBLING_KEYS: List<String> = listOf(
		"dataCategories",
		"features",
		"gvlSpecificationVersion",
		"lastUpdated",
		"purposes",
		"specialFeatures",
		"specialPurposes",
		"stacks",
		"tcfPolicyVersion",
		"vendorListVersion",
	)

	/** [element]'s `vendors` keys, in document order, as numbers. */
	fun vendorKeysOf(document: JsonObject): List<Long> =
		document["vendors"]?.jsonObject?.keys?.map { it.toLong() } ?: emptyList()
}
